GVFS/GVFS.Common/HealthCalculator/EnlistmentHealthCalculator.cs

```diff
@@ -0,0 +1,162 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace GVFS.Common
+{
+    /// <summary>
+    /// Class responsible for the business logic involved in calculating the health statistics
+    /// of a gvfs enlistment. Constructed with the lists of paths for the enlistment, and then
+    /// internally stores the calculated information. Compute or recompute via CalculateStatistics
+    /// with an optional parameter to only look for paths which are under the specified directory
+    /// </summary>
+    public class EnlistmentHealthCalculator
+    {
+        // In the context of this class, hydrated files are placeholders or modified paths
+        // The total number of hydrated files is this.PlaceholderCount + this.ModifiedPathsCount
+        private readonly EnlistmentPathData enlistmentPathData;
+
+        public EnlistmentHealthCalculator(EnlistmentPathData pathData)
+        {
+            this.enlistmentPathData = pathData;
+        }
+
+        public EnlistmentHealthData CalculateStatistics(string parentDirectory)
+        {
+            int gitTrackedItemsCount = 0;
+            int placeholderCount = 0;
+            int modifiedPathsCount = 0;
+            Dictionary<string, int> gitTrackedItemsDirectoryTally = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase);
+            Dictionary<string, int> hydratedFilesDirectoryTally = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase);
+
+            // Parent directory is a path relative to the root of the repository which is already in git format
+            if (!parentDirectory.EndsWith(GVFSConstants.GitPathSeparatorString) && parentDirectory.Length > 0)
+            {
+                parentDirectory += GVFSConstants.GitPathSeparator;
+            }
+
+            if (parentDirectory.StartsWith(GVFSConstants.GitPathSeparatorString))
+            {
+                parentDirectory = parentDirectory.TrimStart(GVFSConstants.GitPathSeparator);
+            }
+
+            gitTrackedItemsCount += this.CategorizePaths(this.enlistmentPathData.GitFolderPaths, gitTrackedItemsDirectoryTally, parentDirectory);
+            gitTrackedItemsCount += this.CategorizePaths(this.enlistmentPathData.GitFilePaths, gitTrackedItemsDirectoryTally, parentDirectory);
+            placeholderCount += this.CategorizePaths(this.enlistmentPathData.PlaceholderFolderPaths, hydratedFilesDirectoryTally, parentDirectory);
+            placeholderCount += this.CategorizePaths(this.enlistmentPathData.PlaceholderFilePaths, hydratedFilesDirectoryTally, parentDirectory);
+            modifiedPathsCount += this.CategorizePaths(this.enlistmentPathData.ModifiedFolderPaths, hydratedFilesDirectoryTally, parentDirectory);
+            modifiedPathsCount += this.CategorizePaths(this.enlistmentPathData.ModifiedFilePaths, hydratedFilesDirectoryTally, parentDirectory);
+
+            Dictionary<string, int> mostHydratedDirectories = new Dictionary<string, int>();

```

Comment:
For consistency, we might want to explicitly include the string comparer. We are not looking up values in the dictionary at the moment, but it is better to specify the correct string comparer in case someone needs to use that functinality in the future.